pkg/agent/cniserver/server.go

```diff
@@ -537,17 +537,20 @@ func (s *CNIServer) cmdDel(_ context.Context, cniConfig *CNIConfig) (*cnipb.CniC
 	if s.isChaining {
 		return s.interceptDel(cniConfig)
 	}
-	// Release IP to IPAM driver
-	if err := ipam.ExecIPAMDelete(cniConfig.CniCmdArgs, cniConfig.K8sArgs, cniConfig.IPAM.Type, infraContainer); err != nil {
-		klog.ErrorS(err, "Failed to delete IP addresses for container", "container", cniConfig.ContainerId)
-		return s.ipamFailureResponse(err), nil
-	}
-	klog.InfoS("Deleted IP addresses for container", "container", cniConfig.ContainerId)
+
 	// Remove host interface and OVS configuration
 	if err := s.podConfigurator.removeInterfaces(cniConfig.ContainerId); err != nil {
 		klog.ErrorS(err, "Failed to remove interfaces for container", "container", cniConfig.ContainerId)
 		return s.configInterfaceFailureResponse(err), nil
 	}
+	klog.InfoS("Deleted interfaces for container", "container", cniConfig.ContainerId)
+
+	// Release IP to IPAM driver
+	if err := ipam.ExecIPAMDelete(cniConfig.CniCmdArgs, cniConfig.K8sArgs, cniConfig.IPAM.Type, infraContainer); err != nil {
```

Comment:
I think one point I was trying to make earlier (in the other PR) is that the main reason we have observed (AFAIK) for a CNI DEL to fail is OVS cleanup. As long as `s.podConfigurator.removeInterfaces` keeps failing, we will not be releasing the Pod IP address, so it will not be available for new Pods.
However, it's probably fair to assume that if `s.podConfigurator.removeInterfaces` keeps failing, then `s.podConfigurator.configureInterfaces` cannot also succeed consistently, so Pod creation is impacted anyway. So releasing the IP address first would not help in this case.

The observation here is that the pool of IPs is much smaller than the pool of OVS resources (e.g. OF ports). But I don't know if it is that relevant. Typically you want to release things in the reverse order compared to allocation, and your point about OVS flows referencing the IP address makes sense.

Reply:
Yes, creating resource should be more demanding than releasing resources. Releasing IPs first wouldn't help much in theory. Besides:
1. The issue didn't happen only when there is something wrong but could also happen when Pod A is being created and Pod B is being deleted, and there is no other available IP.
2. We should avoid creating a case that multiple Pods share the same IP as IP is the identifier in datapath, it may lead to misbehavior to Service and NetworkPolicy. If there is a case cleaning up OVS resources keeps failing, releasing the IP to another Pod could make another Pod be applied this Pod's Policy and used as an irrelavant Service's endpoint wrongly.